emacs-the-best-python-editor/PyEval/pyeval_expression.py

```diff
@@ -0,0 +1,188 @@
+            return self.__output_string
+
+        # Let's start evaluating
+        # Right now, every expression starts with an operand
+        # This is not universally true for functions and parentheses, but we're
+        # not supporting them yet
+        # TODO: Add support for functions and parentheses
+        expecting_operand = True
+
+        # Get the current character to inspect
+        current_char = self.__expr_string[self.__current_position]
+
+        # Loop until we're past the end of the string
+        while (
+            self.__current_position < len(self.__expr_string)
+            and current_char != "$"
+        ):
+
+            # Skip any leading whitespace characters
+            while current_char.isspace():
+                self.__current_position += 1
+                current_char = self.__expr_string[self.__current_position]
+
+            # Store whatever is next in the current_token string
+            current_token = ""
+
+            # If we are looking for an operand
+            if expecting_operand:
+                # First, we need to check for a leading '-' or '+' sign
+                if current_char == "-" or current_char == "+":
+                    current_token += current_char
+                    self.__current_position += 1
+                    current_char = self.__expr_string[self.__current_position]
+
+                # Now we loop for as long as we have numbers
+                while current_char in [
+                    "0",
+                    "1",
+                    "2",
+                    "3",
+                    "4",
+                    "5",
+                    "6",
+                    "7",
+                    "8",
+                    "9",
+                ]:
+                    current_token += current_char
+                    self.__current_position += 1
+                    current_char = self.__expr_string[self.__current_position]
+
+                # We should have a number now - add it to the output string,
+                # space delimited
+                self.__output_string += current_token + " "
+
+                # And after every operand, we need to look for an operator
+                expecting_operand = False
+
+            else:
+                # Here, we just need a single operator, so
+                # Get that operator, validate it, then
+                # Create a new operator object
+                if current_char not in ["+", "-", "*", "/", "%", "^"]:
+                    raise SyntaxError
+
+                current_operator = Operator(current_char)
+
+                # Now comes the shunting yard part
+                # - If the operator stack is empty, push the current operator
+                # - Else
+                #   - While the top of stack operator is higher precedence
+                #     - Pop it and output it.
+                #   - Push the current operator
+
+                if len(self.__operator_stack) == 0:
```

Comment:
You can probably write this as `if not self.__operator_stack:`

Reply:
More Pythonic it is! Change #e68c04c94b839a438c485a4b92b5ecce9d9d70df